nxcomp/Log.h

```diff
@@ -0,0 +1,461 @@
+		return log_level_;
+	}
+
+	void log_level(bool val)
+	{
+		log_level_ = val;
+	}
+
+	bool log_time() const 
+	{
+		return log_time_;
+	}
+
+	void log_time(bool val)
+	{
+		log_time_ = val;
+	}
+
+	bool log_unix_time() const
+	{
+		return log_unix_time_;
+	}
+
+	void log_unix_time(bool val)
+	{
+		log_unix_time_ = val;
+	}
+
+	bool log_location() const
+	{
+		return log_location_;
+	}
+
+	void log_location(bool val)
+	{
+		log_location_ = val;
+	}
+
+	bool log_thread_id() const
+	{
+		return log_thread_id_;
+	}
+
+	void log_thread_id(bool val)
+	{
+		log_thread_id_ = val;
+	}
+
+
+	void flush()
+	{
+		per_thread_data *pdt = get_data();
+		const std::string str = pdt->buffer->str();
+
+		if (!str.empty())
+		{
+			pthread_mutex_lock(&output_lock_);
+			(*stream()) << str;
+			pthread_mutex_unlock(&output_lock_);
+			pdt->buffer->str("");
+		}
+	}
+
+	std::string thread_name() const
+	{
+		return *get_data()->thread_name;
+	}
+
+	void thread_name(std::string str)
+	{
+		*get_data()->thread_name = str;
+	}
+
+	void thread_name(const char *str)
+	{
+		*get_data()->thread_name = str;
+	}
+
+	/**
+	 * True if a message sent to the NXLog object will be sent to the output
+	 *
+	 * This considers two things:
+	 *
+	 * If there's a per-file log level, then it is used
+	 * Otherwise the global log level is used.
+	 *
+	 * If the log level permits the current message to be output, then the
+	 * return value is true.
+	 */ 
+	bool will_log() const;
+	
+
+	/**
+	 * This catches std::flush
+	 */
+	NXLog&  operator<<(std::ostream& (*F)(std::ostream&)) 
+	{
+		if ( will_log() )
+		{
+			if ( synchronized() )
+			{
+				per_thread_data *pdt = get_data();
+				(*pdt->buffer) << F;
+				flush();
+			}
+			else
+			{
+				*(stream()) << F;
+			}
+		}
+
+		return *this; 
+	} 
+
+	template<typename T>
+	friend NXLog& operator<<(NXLog& out, const T& value);
+
+	friend NXLog& operator<< (NXLog& out, const NXLogStamp& value);
+};
+
+
+extern NXLog nx_log;
+
+
+#define nxstamp(l) NXLogStamp(__FILE__, __func__, __LINE__, l)
+
+
+#define nxdbg    nx_log << nxstamp(NXDEBUG)
+#define nxinfo   nx_log << nxstamp(NXINFO)
+#define nxwarn   nx_log << nxstamp(NXWARNING)
+#define nxerr    nx_log << nxstamp(NXERROR)
+#define nxfatal  nx_log << nxstamp(NXFATAL)
+
+
+NXLog& operator<< (NXLog& out, const NXLogStamp& value);
+
+
+template <typename T>
+bool has_newline(T value)
+{
+	return false;
+}
+
+template <char*>
+static bool has_newline(char *value)
+{
+	return strstr(value, "\n") != NULL;
+}
+
+template <char> 
+static bool has_newline(char value)
+{
+	return value == '\n';
+}
+
+template <std::string&>
+static bool has_newline(std::string &value)
+{
+	return value.find_first_of("\n") != std::string::npos;
+}
+
+static size_t ss_length(std::stringstream *ss)
+{
+	size_t pos = ss->tellg();
+	size_t ret = 0;
+	ss->seekg(0, std::ios::end);
+	ret = ss->tellg();
+	ss->seekg(pos, std::ios::beg);
+	return ret;
+}
+
+template <typename T>
+NXLog& operator<<(NXLog& out, const T& value)
```

Comment:
This code is duplicated three times and - worse - each time it's behaving a it differently (flushed at the beginning, flushed at the end, ...)

The non-templated function taking an NSLogStamp value should at least be using the templated version after it set up its stuff... and why can't the templated version use the operator<< version from NXLog again? (Serious question, I might be overlooking something crucial.)

Also, we should make sure that the last message is flushed when the NXLog object is deleted, otherwise it will never make it through to a log file or stream.